unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+
+        match flat.data.Type {
+            UnixFsType::Directory => {
+                let inner = InnerEntry::new_root_dir(metadata);
+
+                let links = flat.links
+                    .into_iter()
+                    .enumerate()
+                    // 2 == number of ancestors this link needs to have on the path, this is after
+                    // some trial and error so not entirely sure why ... ancestors always includes
+                    // the empty root in our case.
+                    .map(|(nth, link)| convert_link(2, nth, link));
+
+                Self::walk_directory(links, inner)
+            },
+            UnixFsType::HAMTShard => {
+                let inner = InnerEntry::new_root_bucket(metadata);
+
+                let links = flat.links
+                    .into_iter()
+                    .enumerate()
+                    // 1 == ancestors should be only the [""]
+                    .map(|(nth, link)| convert_sharded_link(1, nth, link));
+
+                Self::walk_directory(links, inner)
+            },
+            UnixFsType::Raw | UnixFsType::File => {
+                let (bytes, metadata, step) = IdleFileVisit::default()
+                    .start_from_parsed(flat, cache)?;
+
+                if let Some(visit) = step {
+                    Ok(Walk::MultiBlockFile {
+                        metadata,
+                        visit
+                    })
+                } else {
+                    Ok(Walk::SingleBlockFile {
+                        bytes,
+                        metadata
+                    })
+                }
+            },
+            UnixFsType::Metadata => todo!("metadata?"),
+            UnixFsType::Symlink => todo!("?"),
+        }
+    }
+
+    fn walk_directory<'a, I>(mut links: I, current: InnerEntry) -> Result<Walk<'a>, Error>
+        where I: Iterator<Item = Result<(Cid, String, usize), InvalidCidInLink>> + 'a,
+    {
+        if let Some(next) = links.next() {
+            let next = Some(next?);
+            let pending = links.collect::<Result<Vec<_>, _>>()?;
+
+            Ok(Walk::Walker(Walker {
+                current,
+                next,
+                pending,
+            }))
+        } else {
+            Ok(Walk::EmptyDirectory {
+                metadata: current.into(),
+            })
+        }
+    }
+
+    pub fn as_entry<'a>(&'a self) -> Entry<'a> {
+        self.current.as_entry()
+    }
+
+    pub fn pending_links(&self) -> (&Cid, impl Iterator<Item = &Cid>) {
+        use InnerKind::*;
+        let cids = self.pending.iter().map(|(cid, ..)| cid).rev();
+        match self.current.kind {
+            File(_, Some(ref visit)) => {
+                let (first, rest) = visit.pending_links();
+                (first, Either::Left(rest.chain(cids)))
+            },
+            _ => {
+                let next = self.next.as_ref()
+                    .expect("validated in start and continue_walk we have the next");
```

Comment:
This surely will need more comments and this message is a bit wrong already.

When we have an active `FileVisit` (like in the previous match arm) we might not actually have anything in the `self.next`, which is also missing from the iterator above! (I went back and forth with `Vec<(...)>` and `next` and `Vec<(...)>`.)